File: backend/app/services/backup_normalizer.py
import hashlib
import re

VOLATILE_HEADER_RE = re.compile(
    r"^#\s+\d{4}-\d{2}-\d{2}\s+\d{2}:\d{2}:\d{2}\s+by\s+RouterOS\b.*$",
    re.MULTILINE,
)


def normalize_rsc(rsc_text: str) -> str:
    """Normalize RouterOS .rsc script export by removing volatile timestamp headers

    and standardizing newlines and trailing whitespace.
    """
    if not rsc_text:
        return ""
    # Standardize line endings
    text = rsc_text.replace("\r\n", "\n").replace("\r", "\n")
    # Strip volatile RouterOS timestamp line
    text = VOLATILE_HEADER_RE.sub("", text)
    # Strip trailing whitespaces per line
    lines = [line.rstrip() for line in text.split("\n")]
    # Remove leading blank lines
    while lines and not lines[0]:
        lines.pop(0)
    # Remove trailing blank lines
    while lines and not lines[-1]:
        lines.pop()
    return "\n".join(lines)


def compute_fingerprint(rsc_text: str) -> str:
    """Return SHA-256 hex digest of normalized RouterOS configuration script."""
    normalized = normalize_rsc(rsc_text)
    return hashlib.sha256(normalized.encode("utf-8")).hexdigest()

File: backend-go/internal/services/backup_normalizer.go
package services

import (
	"crypto/sha256"
	"encoding/hex"
	"regexp"
	"strings"
)

var volatileHeaderRegex = regexp.MustCompile(`(?m)^#\s+\d{4}-\d{2}-\d{2}\s+\d{2}:\d{2}:\d{2}\s+by\s+RouterOS\b.*$`)

// NormalizeRSC strips volatile timestamp headers and normalizes line endings and whitespace.
func NormalizeRSC(rscText string) string {
	if rscText == "" {
		return ""
	}

	// Standardize line endings
	text := strings.ReplaceAll(rscText, "\r\n", "\n")
	text = strings.ReplaceAll(text, "\r", "\n")

	// Strip volatile RouterOS timestamp header
	text = volatileHeaderRegex.ReplaceAllString(text, "")

	// Strip trailing whitespace per line
	lines := strings.Split(text, "\n")
	for i, l := range lines {
		lines[i] = strings.TrimRight(l, " \t")
	}

	// Remove leading empty lines
	for len(lines) > 0 && lines[0] == "" {
		lines = lines[1:]
	}

	// Remove trailing empty lines
	for len(lines) > 0 && lines[len(lines)-1] == "" {
		lines = lines[:len(lines)-1]
	}

	return strings.Join(lines, "\n")
}

// ComputeFingerprint returns the SHA-256 hex digest of the normalized RouterOS configuration.
func ComputeFingerprint(rscText string) string {
	normalized := NormalizeRSC(rscText)
	hash := sha256.Sum256([]byte(normalized))
	return hex.EncodeToString(hash[:])
}

